mdbook-spec/src/grammar.rs

```diff
@@ -0,0 +1,389 @@
+            | ExpressionKind::Prose(_)
+            | ExpressionKind::Break(_)
+            | ExpressionKind::Unicode(_) => {}
+            ExpressionKind::Charset(set) => {
+                for ch in set {
+                    match ch {
+                        Characters::Named(s) => callback(s),
+                        Characters::Terminal(_) | Characters::Range(_, _) => {}
+                    }
+                }
+            }
+        }
+    }
+
+    fn is_break(&self) -> bool {
+        matches!(self.kind, ExpressionKind::Break(_))
+    }
+}
+
+static GRAMMAR_RE: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"(?ms)^```grammar,([^\n]+)\n(.*?)^```").unwrap());
+static NAMES_RE: LazyLock<Regex> =
+    LazyLock::new(|| Regex::new(r"(?m)^([A-Za-z0-9_]+)(?: \([^)]+\))? ->").unwrap());
+
+/// Loads the [`Grammar`] from the book.
+pub fn load_grammar(book: &Book, diag: &mut Diagnostics) -> Grammar {
+    let mut grammar = Grammar::default();
+    for item in book.iter() {
+        let BookItem::Chapter(ch) = item else {
+            continue;
+        };
+        if ch.is_draft_chapter() {
+            continue;
+        }
+        let path = ch.path.as_ref().unwrap().to_owned();
+        for cap in GRAMMAR_RE.captures_iter(&ch.content) {
+            let category = &cap[1];
+            let input = &cap[2];
+            if let Err(e) = parser::parse_grammar(input, &mut grammar, category, &path) {
+                warn_or_err!(diag, "failed to parse grammar in {path:?}: {e}");
+            }
+        }
+    }
+    check_undefined_nt(&grammar, diag);
+    check_unexpected_roots(&grammar, diag);
+    grammar
+}
+
+/// Checks for nonterminals that are used but not defined.
+fn check_undefined_nt(grammar: &Grammar, diag: &mut Diagnostics) {
+    grammar.visit_nt(&mut |nt| {
+        if !grammar.productions.contains_key(nt) {
+            warn_or_err!(diag, "non-terminal `{nt}` is used but not defined");
+        }
+    });
+}
+
+/// This checks that all the grammar roots are what we expect.
+///
+/// This is intended to help catch any unexpected misspellings, orphaned
+/// productions, or general mistakes.
+fn check_unexpected_roots(grammar: &Grammar, diag: &mut Diagnostics) {
+    let mut set: HashSet<_> = grammar.name_order.iter().map(|s| s.as_str()).collect();
+    grammar.visit_nt(&mut |nt| {
+        set.remove(nt);
+    });
+    // TODO: We may want to rethink how some of these are structured.
+    let expected: HashSet<_> = [
+        "CfgAttrAttribute",
+        "CfgAttribute",
+        "CHAR",
+        "Crate",
+        "INNER_LINE_DOC",
+        "LINE_COMMENT",
+        "MetaListIdents",
+        "MetaListNameValueStr",
+        "MetaListPaths",
+        "MetaWord",
+        "OUTER_LINE_DOC",
+    ]
+    .into_iter()
+    .collect();
+    if set != expected {
+        let new: Vec<_> = set.symmetric_difference(&expected).collect();
+        let removed: Vec<_> = expected.symmetric_difference(&set).collect();
```

Comment:
Actually, this is wrong.  As the name implies, the "symmetric" difference is symmetric, so these two lists are always the same.

Reply:
Fixed in a commit.

Reply:
Oops, yea, just picked the wrong function.